PEP 831: Frame pointers (#4896)
python/peps: peps/pep-0831.rst

Author: Pablo Galindo Salgado <pablogsal@python.org>,
        Ken Jin <kenjin@python.org>,
        Savannah Ostrowski <savannah@python.org>,
Discussions-To: Pending
Status: Draft
Type: Standards Track
Created: 14-Mar-2026
Python-Version: 3.15
Post-History: Pending

Abstract
========

This PEP proposes two things:

1. **Build CPython with frame pointers by default on platforms that support
   them.**  The default build configuration is changed to compile the
   interpreter with ``-fno-omit-frame-pointer`` and
   ``-mno-omit-leaf-frame-pointer``. The flags are added to ``CFLAGS``, so they
   apply to the interpreter itself and propagate to C extension modules built
   against this Python via ``sysconfig``.  An opt-out ``configure`` flag
   (``--without-frame-pointers``) is provided for deployments that require
   maximum raw throughput.

2. **Strongly recommend that all build systems in the Python ecosystem build
   with frame pointers by default.**  This PEP recommends that *every* compiled
   component that participates in the Python call stack (C extensions, Rust
   extensions, embedding applications, and native libraries) should enable

>    frame pointers.  A frame-pointer chain is only as complete as its weakest

Reviewer: ```suggestion
   frame pointers.  A frame-pointer chain is only as strong as its weakest
```

Author: Done, thanks.

   link: a single library without frame pointers breaks profiling, debugging,
   and tracing for the entire process.

Frame pointers are a CPU register convention that allows profilers, debuggers,
and system tracing tools to reconstruct the call stack of a running process
quickly and reliably.  Omitting them (the compiler's default at ``-O1`` and
above) prevents these tools from producing useful call stacks for Python
processes, and undermines the perf trampoline support CPython shipped in 3.12.

> The measured overhead is under 2% geometric mean for typical workloads.

Reviewer: For what platform, build configuration, etc, etc?
Can you link to the data, otherwise the readers (at least this reader's) first thought is "where's the evidence?"

Author: Added a cross-reference to the Backwards Compatibility section which has the full per-platform table (8 machines across arm64 and x86-64, geometric mean across 80 pyperformance benchmarks). The abstract is intentionally concise but now points readers to the evidence.

Multiple major Linux distributions, language runtimes, and Python ecosystem
tools have already adopted this change.  No existing PEP covers this topic;
CPython issue `#96174`_ has been open since August 2022 without resolution.

Motivation
==========

Python's observability story (profiling, debugging, and system-level tracing)

> is fundamentally limited by the absence of frame pointers. The core motivation

Reviewer: This isn't quite true. It isn't impossible without frame pointers, just much, much harder.

Author: The text says "fundamentally limited", not impossible, no?

of this PEP is to make Python observable by default, so that profilers are
faster and more accurate, debuggers are more reliable, and eBPF-based tools
are functional without workarounds.

Today, users who want to profile CPython with system tools must rebuild the
interpreter with special compiler flags, a step that most users cannot or will
not take.  The Fedora 38 frame-pointer proposal [#fedora38]_ highlights this as
the key problem: without frame pointers in the default build, developers must
"recompile their program with sufficient debugging information" and "reproduce
the scenario under which the software performed poorly," which is often
impossible for production issues.  Ubuntu 24.04's analysis [#ubuntu2404]_ makes
the same argument: frame pointers "allow bcc-tools, bpftrace, perf and other
such tooling to work out of the box."  The goal of this PEP is to make that the
default experience for Python.

The performance wins that profiling enables far outweigh the modest overhead of
frame pointers.  As Brendan Gregg notes: "I've seen frame pointers help find
performance wins ranging from 5% to 500%" [#gregg2024]_.  A 0.5-2% overhead that

> unlocks the ability to find 5-500% improvements is a favourable trade.

Reviewer: Make it clear, that large speedups are not going to happen for CPython, but it might happen for larger systems using Python as a glue language.

Author: Good catch. Clarified that the 5-500% wins Gregg describes come from identifying hot paths in production systems, not from enabling frame pointers on CPython itself.

What Are Frame Pointers?
------------------------

When a program runs, each function call creates a **stack frame**, a block of
memory on the call stack that holds the function's local variables, its
arguments, and the address to return to when the function finishes.  The **call
stack** is the chain of all active stack frames: it records which function
called which, all the way from ``main()`` to the function currently executing.

> A **frame pointer** is a CPU register (``%rbp`` on x86-64, ``x29`` on AArch64)

Reviewer: ```suggestion
A **frame pointer** is a CPU register (for example,``%rbp`` on x86-64, ``x29`` on AArch64)
```
Or use a footnote to give a longer list?

Author: Added "for example,". A footnote with more architectures could work but might be overkill since x86-64 and AArch64 cover the vast majority of Python deployments. Happy to add one if you think it would help.

that each function sets to point to the base of its own stack frame.  Each
frame also stores the *previous* frame pointer, creating a linked list through
the entire call stack::

>     ┌──────────────────┐

Reviewer: Have the arrows actually point to the previous frame. I think it is OK to use an image,instead of ASCII art.

Author: Yeah perhaps an image may be better.

Just to clarify: The arrows do point to the previous frame as  each saved %rbp ──► points to the frame above it (i.e. its caller), which is the correct direction for a frame-pointer chain. The chain walks up (toward main). I can look into replacing with an image if the ASCII art is unclear, but it would complicate the diff/review cycle. Open to suggestions.

    │  main()          │ ◄─── frame pointer chain
    │  saved %rbp ─────┼──► (bottom of stack)
    ├──────────────────┤
    │  PyRun_String()  │
    │  saved %rbp ─────┼──► main's frame
    ├──────────────────┤
    │  _PyEval_Eval…() │
    │  saved %rbp ─────┼──► PyRun_String's frame
    ├──────────────────┤
    │  call_function() │ ◄─── current %rbp
    │  saved %rbp ─────┼──► _PyEval_Eval's frame
    └──────────────────┘

> Stack unwinding is the process of walking this chain to reconstruct the call

Reviewer: Is "unwinding" the correct term?
Isn't stack unwinding what happens when a C++ (or similar) exception is thrown?
Unwinding sounds destructive.

The term "stack walking" is used in the GC handbook, I've also heard "traversal" used.

Author: "Stack unwinding" is the standard term used by DWARF (the spec defines "Call Frame Information" for unwinding), perf (--call-graph modes are documented as "unwinding"), libunwind (the name itself), GDB internals, and the Linux kernel's ORC unwinder. It predates C++ exceptions: C++ co-opted the term for its destructors-during-exception-propagation mechanism, but the original meaning reconstructing the call chain) is the primary one in the profiling/debugging world. I'd  to keep the standard terminology here.

stack.  Profilers do it to find out where the program is spending time;
debuggers do it to show backtraces; crash handlers do it to produce useful

> error reports.  With frame pointers, unwinding is a simple pointer chase: read

Reviewer: ```suggestion
error reports.  With frame pointers, unwinding is a simply following pointers: read
```

I don't know if you can expect the readership to known the term "pointer chasing"

Author: Done, good point about "pointer chasing" being jargon.

> ``%rbp``, follow the link, repeat. It takes microseconds and requires no

Reviewer: ```suggestion
``%rbp``, follow the link, repeat. It is very fast and requires no
```
How long it actually takes depends on *many* factors.

Author: Removed the timing claim from this introductory section. The precise numbers (116 ns for 64 frames) appear later in the benchmarks section where they have proper context.

external data.

At optimisation levels ``-O1`` and above, GCC and Clang omit frame pointers by
default [#gcc_fomit]_.  This frees the ``%rbp`` register for general use,
giving the optimiser one more register to work with. On x86-64 this is a gain
of one register out of 16 (about 7%).  The performance benefit is small

> (typically 0.5-2%) but it was considered worthwhile when the convention was

Reviewer: Again, you need to be more nuanced about how much things slow down, and for which platforms.

I'd use terms like "a few" rather than 0.5-2 and add a link to a section with a more detailed breakdown

Author: Changed to "typically a few percent" and added a cross-reference to the detailed analysis section which has the full platform-by-platform breakdown.

established for 32-bit x86, where the gain was one register out of 6 (~20%).

Without frame pointers, the linked list does not exist.  Tools that need to

> walk the call stack must instead parse DWARF debug information (a complex,

Reviewer: Don't forget Windows. It would make your case more compelling, as tools would need to parse DWARF *and* understand what Windows does.

Author: Good point. Added a mention of Windows `.pdata` / `.xdata` unwind metadata alongside DWARF.

       |
       ---_PyEval_EvalFrameDefault
          (truncated, no further frames available)

The same profile **with** frame pointers shows the full chain::

    100.00%  python  libpython3.14.so  _PyEval_EvalFrameDefault
       |
       ---_PyEval_EvalFrameDefault
          |
          +--PyObject_Call
          |  PyRun_StringFlags
          |  PyRun_SimpleStringFlags
          |  Py_RunMain
          |  main
          |
          +--call_function
             fast_function
             ...

The first trace is useless for diagnosing performance problems; the second
tells the developer exactly what code path is hot.

Profilers Are Slower and Less Accurate Without Frame Pointers
-------------------------------------------------------------

Statistical profilers (``perf``, py-spy, Austin, Pyroscope, Parca, and others)
work by periodically sampling the call stack of a running process.  With frame
pointers, this sampling is a simple pointer chase: the profiler reads ``%rbp``,
follows the chain, and reconstructs the full call stack in microseconds.  This
is fast enough to sample at 10,000 Hz in production with negligible overhead.

> Without frame pointers, the profiler must fall back to DWARF unwinding, a

Reviewer: Didn't you just say this a few paragraphs earlier?
https://github.com/python/peps/pull/4896/changes#diff-cce3082fa07cd352af453f46a4e6f23b15645c79d58cb902da789f5c463cc92aR115

Author: The earlier mention (in "What Are Frame Pointers?") is a brief general overview; this section ("Profilers Are Slower...") dives into the specific profiler impact with data (55x larger perf.data, 210x slower unwinding, etc.). Some overlap is intentional as the first establishes the concept, the second builds the quantitative case. I could trim the first mention further if you think it's too redundant.

``bpf_get_stackid()``, these implementations parse ``.eh_frame`` sections in
userspace, convert them to compact stack-delta lookup tables, load those tables
into BPF maps, and then evaluate the tables in a custom BPF program that
manually reads stack memory with ``bpf_probe_read_user()``.  This requires 500+
lines of BPF code (compared to fewer than 50 for the frame-pointer path),
demands per-process startup overhead to parse and load debug info, consumes
significant BPF map memory for the stack-delta tables, and is cutting-edge
vendor-specific infrastructure not available in standard tooling such as
bpftrace, bcc, or ``perf`` [#polarsignals]_.

For the vast majority of eBPF use cases (bpftrace one-liners, bcc tools, custom
BPF programs for production monitoring), frame pointers are the only viable
unwinding mechanism because they are the only mechanism the kernel's built-in
helpers support.

CPython's own documentation already states the recommended fix::

    For best results, Python should be compiled with
    CFLAGS='-fno-omit-frame-pointer -mno-omit-leaf-frame-pointer'
    as this allows profilers to unwind using only the frame pointer
    and not on DWARF debug information.

>     -- docs.python.org/3/howto/perf_profiling.html

Reviewer 2: Fix link, no code formatting for quoting text:

```suggestion
:ref:`CPython's own documentation <python:perf_profiling>`__ already states the
recommended fix:

    For best results, Python should be compiled with
    ``CFLAGS="-fno-omit-frame-pointer -mno-omit-leaf-frame-pointer"``
    as this allows profilers to unwind using only the frame pointer
    and not on DWARF debug information.
```

generates small JIT stubs that make Python function names visible to ``perf``.
These stubs contain no DWARF information; the only way a profiler can walk
through them is via the frame-pointer chain.  The ``-Xperf`` feature as shipped
in 3.12 therefore produces broken stacks on any installation that was not
explicitly rebuilt with ``-fno-omit-frame-pointer``.

Python 3.13 added ``-Xperf_jit``, a DWARF-based alternative, but it requires
``perf`` >= 6.8, produces substantially larger data files than the
frame-pointer path, and is not suitable for continuous production profiling due
to its per-sample overhead.

Distributions Are Waiting for Upstream
--------------------------------------

Fedora 38 [#fedora38]_, Ubuntu 24.04 LTS [#ubuntu2404]_, and Arch Linux have
all rebuilt their entire package trees with ``-fno-omit-frame-pointer
-mno-omit-leaf-frame-pointer``.  However, Ubuntu 24.04 LTS explicitly exempted
CPython::

    In cases where the impact is high (such as the Python
    interpreter), we'll continue to omit frame pointers until
    this is addressed.

    -- ubuntu.com/blog/ubuntu-performance-engineering-with-

>        frame-pointers-by-default

Reviewer 2: ```suggestion
-mno-omit-leaf-frame-pointer``.  However, Ubuntu 24.04 LTS
`explicitly exempted CPython
<https://ubuntu.com/blog/ubuntu-performance-engineering-with-frame-pointers-by-default>`__:

    In cases where the impact is high (such as the Python
    interpreter), we'll continue to omit frame pointers until
    this is addressed.
```

python.org, build from source, use pyenv, or work on Debian, RHEL, openSUSE, or
any other distribution that has not adopted frame pointers system-wide get no
frame pointers regardless of what their local distribution does. An upstream
default resolves this permanently.

Mixed Python/C Profiling Requires a Continuous Chain
----------------------------------------------------

Real Python applications spend substantial time in C extension modules: NumPy,
cryptographic libraries, compression, database drivers, and so on.  For a
``perf`` flame graph to show the full path from Python code through a C
extension and into a system library, the frame-pointer chain must be continuous
through the interpreter **and** through every extension in the call stack.  A
gap at any point, whether in ``_PyEval_EvalFrameDefault`` or in a C extension,
breaks the chain for the entire process.

The need for a continuous chain is precisely why the flags must propagate to
extension builds. If only the interpreter has frame pointers but extensions do
not, the chain is still broken at every C extension boundary.  By adding the
flags to ``CFLAGS`` as reported by ``sysconfig``, extension builds that consume

> CPython's compiler flags (for example via ``pip install``, ``setuptools``, or

Reviewer 2: ```suggestion
CPython's compiler flags (for example via ``pip install``, Setuptools, or
```

tail-call interpreter [#pbs992]_.  These numbers are consistent with this PEP's
first-party measurements and with Fedora/Ubuntu data.

Major Linux distributions (Fedora 38 [#fedora38]_, Ubuntu 24.04 [#ubuntu2404]_,
Arch Linux) have rebuilt their entire package trees with frame pointers.
PyTorch, Node.js, Redis, Go, and .NET have all adopted frame pointers in their
default builds (see `Industry Consensus Has Shifted Decisively`_ in the
Rationale for the full list).

An upstream default aligns CPython with the reality that the ecosystem has
already adopted.

Specification
=============

Build System Changes
--------------------

The following changes are made to ``configure.ac``::

    AX_CHECK_COMPILE_FLAG([-fno-omit-frame-pointer],
      [CFLAGS="$CFLAGS -fno-omit-frame-pointer"])
    AX_CHECK_COMPILE_FLAG([-mno-omit-leaf-frame-pointer],
      [CFLAGS="$CFLAGS -mno-omit-leaf-frame-pointer"])

Using ``CFLAGS`` ensures:

1. The flags apply to all ``*.c`` files compiled as part of the interpreter:
   the ``python`` binary, ``libpython``, and built-in extension modules under
   ``Modules/``.
2. The flags **are** written into the ``sysconfig`` data, so that third-party C

>    extensions built against this Python (via ``pip``, ``setuptools``, or direct

Reviewer 2: ```suggestion
   extensions built against this Python (via pip, Setuptools, or direct
```

functions.  Without it, the compiler may drop the frame pointer in any function
that makes no further calls, even when ``-fno-omit-frame-pointer`` is set.
Fedora, Ubuntu, and Arch Linux all include this flag; it ensures a profiler
sampling inside a leaf function still recovers a complete call chain.

Opt-Out Configure Flag
----------------------

A new ``configure`` option is added::

    --without-frame-pointers

When specified, neither flag is added to ``CFLAGS``.  This is appropriate for
deployments that have measured an unacceptable regression on their specific
workload, or for distributions that inject frame-pointer flags at a higher
level and wish to avoid double-specification, analogous to Fedora's per-package
``%undefine _include_frame_pointers`` macro.

Extension authors who wish to override the default for a specific module can
pass ``-fomit-frame-pointer`` in their ``extra_compile_args`` or via
environment variables; the last flag on the command line wins under GCC and
Clang.

Ecosystem Impact
----------------

Because the flags are in ``CFLAGS``, they propagate automatically to consumers
that build against CPython's reported compiler flags, such as C extensions

> built via ``pip``, ``setuptools``, or direct ``sysconfig`` queries.  Those

Reviewer 3: nit: I don't think pip directly builds any C extension, build backends do that

Reviewer 2: ```suggestion
built via pip, Setuptools, or direct ``sysconfig`` queries.  Those
```

Frame-pointer unwinding remains the only method that is simultaneously:
kernel-side, async-signal-safe, BPF-compatible, depth-unlimited, and produces
compact profiling data.  It is the method that works everywhere with no
additional configuration.

Backwards Compatibility
=======================

Binary Compatibility
--------------------

Enabling frame pointers does not change the Python ABI.  The stable ABI
(:pep:`384`), the limited C API, and the ``PyObject`` memory layout are all
unaffected.  No compiled extension module will fail to load or behave
incorrectly.

Performance
-----------

Full pyperformance results comparing the frame-pointer build against an
identical build without frame pointers (geometric mean across 80
benchmarks [#missing_benchmarks]_). For reproducibility,
pyperformance JSON files can be found in
[#benchmarks]_. Benchmark visualization can be found in the Appendix:

=====================================  =======================
Machine                                Geometric mean overhead
=====================================  =======================
Apple M2 Mac Mini (arm64)              1.006x slower
macOS M3 Pro (arm64)                   1.001x slower

> Raspberry Pi (aarch64).                1.002x slower

Reviewer 2: ```suggestion
Raspberry Pi (aarch64)                 1.002x slower
```

.. [#almalinux_fp] AlmaLinux, "Introducing AlmaLinux OS Kitten 10",
   October 2024.  "We are re-enabling [frame pointers]."
   https://almalinux.org/blog/2024-10-22-introducing-almalinux-os-kitten/

.. [#perforator] Yandex, "Perforator: cluster-wide continuous
   profiling tool", open-sourced January 2025.
   https://github.com/yandex/perforator

.. [#pyroscope] Grafana Pyroscope, "eBPF profiling troubleshooting",
   Grafana documentation.
   https://grafana.com/docs/pyroscope/latest/configure-client/grafana-alloy/ebpf/troubleshooting/

.. [#libunwind] libunwind, "A portable and efficient C programming
   interface to determine the call-chain of a program".
   https://www.nongnu.org/libunwind/

.. [#framehop] framehop, "Stack frame unwinding for profilers",
   by Markus Stange.
   https://github.com/mstange/framehop

.. [#samply] samply, "A command-line sampling profiler for macOS
   and Linux", by Markus Stange.
   https://github.com/mstange/samply

.. [#benchmarks] Python frame-pointer benchmark pyperformance JSON files
   https://github.com/Fidget-Spinner/python-framepointer-bench

.. [#missing_benchmarks] Some benchmarks are missing due to incompatibilities with

>    Python 3.15alpha.

Reviewer 2: ```suggestion
   Python 3.15 alpha.
```

.. _#96174: https://github.com/python/cpython/issues/96174
.. _python/cpython issue #96174: https://github.com/python/cpython/issues/96174
.. _#126910: https://github.com/python/cpython/issues/126910

Appendix
========

For all graphs below, the green dots are geometric means of the
individual benchmark's median, while orange lines are the median of our data points.

> Hollow circles reperesent outliers.

Reviewer 2: ```suggestion
Hollow circles represent outliers.
```

The first graph is the overall effect on pyperformance seen on each system.
Apart from the Ubuntu AWS Graviton System, all system configurations have below 2%
geometric mean and median slowdown:

.. image:: pep-0831_perf_over_baseline.svg

For individual benchmark results, see the following:

> .. image:: pep-0831_perf_over_baseline_indiv.svg

Reviewer 2: Please add alt text.

Reviewer 4: I agree. A side note: I'm surprised that screen readers are not able to read svg? Or perhaps I should try embedding a PDF for accessibility reasons?

Reviewer 2: Aha, you can add `<title>` and `<desc>` inside the SVG:

https://www.a11y-collective.com/blog/svg-accessibility/

Alt might be easier though.

Accessibility of PDFs is a whole other thing, and it's usually easier to manage SVGs in the repo than PDFs.

Reviewer 4: done
